components/layout_2020/flow/mod.rs

```diff
@@ -479,3 +470,104 @@ fn layout_in_flow_non_replaced_block_level<'a>(
+            (inline, block)
+        },
+        (LengthOrAuto::LengthPercentage(inline), LengthOrAuto::Auto) => {
+            (inline, inline / intrinsic_ratio)
+        },
+        (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(block)) => {
+            (block * intrinsic_ratio, block)
+        },
+        (LengthOrAuto::Auto, LengthOrAuto::Auto) => (intrinsic_size.inline, intrinsic_size.block),
+    };
+    let (margin_inline_start, margin_inline_end) = solve_inline_margins_for_in_flow_block_level(
+        containing_block,
+        pb.inline_sum(),
+        computed_margin.inline_start,
+        computed_margin.inline_end,
+        inline_size,
+    );
+    let margin = Sides {
+        inline_start: margin_inline_start,
+        inline_end: margin_inline_end,
+        block_start: computed_margin.block_start.auto_is(Length::zero),
+        block_end: computed_margin.block_end.auto_is(Length::zero),
+    };
+    let containing_block_for_children = ContainingBlock {
+        inline_size,
+        block_size: LengthOrAuto::LengthPercentage(block_size),
+        mode,
+    };
+    // https://drafts.csswg.org/css-writing-modes/#orthogonal-flows
+    assert_eq!(
+        containing_block.mode, containing_block_for_children.mode,
+        "Mixed writing modes are not supported yet"
+    );
+    let independent_layout = replaced.layout(layout_context, style, &containing_block_for_children);
+    let relative_adjustement = relative_adjustement(
+        style,
+        inline_size,
+        LengthOrAuto::LengthPercentage(block_size),
+    );
+    let content_rect = Rect {
+        start_corner: Vec2 {
+            block: pb.block_start + relative_adjustement.block,
+            inline: pb.inline_start + relative_adjustement.inline + margin.inline_start,
+        },
+        size: Vec2 {
+            block: block_size,
+            inline: inline_size,
+        },
+    };
+    BoxFragment {
+        style: style.clone(),
+        children: independent_layout.fragments,
+        content_rect,
+        padding,
+        border,
+        block_margins_collapsed_with_children: CollapsedBlockMargins::from_margin(&margin),
+        margin,
+    }
+}
+
+fn solve_inline_margins_for_in_flow_block_level(
+    containing_block: &ContainingBlock,
+    padding_border_inline_sum: Length,
+    computed_margin_inline_start: LengthOrAuto,
+    computed_margin_inline_end: LengthOrAuto,
+    inline_size: Length,
+) -> (Length, Length) {
+    let inline_margins = containing_block.inline_size - padding_border_inline_sum - inline_size;
+    match (computed_margin_inline_start, computed_margin_inline_end) {
+        (LengthOrAuto::Auto, LengthOrAuto::Auto) => (inline_margins / 2., inline_margins / 2.),
+        (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(end)) => (inline_margins - end, end),
+        (LengthOrAuto::LengthPercentage(start), _) => (start, inline_margins - start),
```

Comment:
Nit: This `_` can only be `Auto`, right? Could we make it a specific pattern?

Reply:
It can be `LengthOrAuto::LengthPercentage` too, which we ignore because that's over-constraining.